src/extensions/idle.rs

```diff
@@ -93,16 +93,26 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
     fn wait_inner(&mut self) -> Result<()> {
         let mut v = Vec::new();
-        match self.session.readline(&mut v).map(|_| ()) {
-            Err(Error::Io(ref e))
-                if e.kind() == io::ErrorKind::TimedOut || e.kind() == io::ErrorKind::WouldBlock =>
-            {
-                // we need to refresh the IDLE connection
-                self.terminate()?;
-                self.init()?;
-                self.wait_inner()
+        loop {
+            let res = match self.session.readline(&mut v).map(|_| ()) {
+                Err(Error::Io(ref e))
+                    if e.kind() == io::ErrorKind::TimedOut
+                        || e.kind() == io::ErrorKind::WouldBlock =>
+                {
+                    // we need to refresh the IDLE connection
+                    self.terminate()?;
+                    self.init()?;
+                    self.wait_inner()
```

Comment:
I think this should probably be a `return self.wait_inner();`. At least then there's a hope that this might get tail-recursion optimized.

Reply:
Thanks for pointing this out! I am changing the recursive path here.